File: app/src/main/java/com/zxcx/zhizhe/ui/loginAndRegister/channelRegister/ChannelRegisterActivity.java
package com.zxcx.zhizhe.ui.loginAndRegister.channelRegister;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.meituan.android.walle.WalleChannelReader;
import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.event.PhoneConfirmEvent;
import com.zxcx.zhizhe.event.RegisterEvent;
import com.zxcx.zhizhe.mvpBase.MvpActivity;
import com.zxcx.zhizhe.ui.loginAndRegister.forget.ForgetPasswordContract;
import com.zxcx.zhizhe.ui.loginAndRegister.forget.ForgetPasswordPresenter;
import com.zxcx.zhizhe.ui.loginAndRegister.forget.PhoneUnRegisteredDialog;
import com.zxcx.zhizhe.ui.loginAndRegister.login.LoginBean;
import com.zxcx.zhizhe.ui.loginAndRegister.register.PhoneConfirmDialog;
import com.zxcx.zhizhe.ui.loginAndRegister.register.SMSCodeVerificationBean;
import com.zxcx.zhizhe.utils.Constants;
import com.zxcx.zhizhe.utils.MD5Utils;
import com.zxcx.zhizhe.utils.Utils;
import com.zxcx.zhizhe.utils.ZhiZheUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import cn.jpush.android.api.JPushInterface;
import cn.smssdk.EventHandler;
import cn.smssdk.SMSSDK;

public class ChannelRegisterActivity extends MvpActivity<ForgetPasswordPresenter> implements ForgetPasswordContract.View {

    @BindView(R.id.iv_forget_close)
    ImageView mIvForgetClose;
    @BindView(R.id.et_forget_phone)
    EditText mEtForgetPhone;
    @BindView(R.id.iv_forget_phone_clear)
    ImageView mIvForgetPhoneClear;
    @BindView(R.id.et_forget_verification_code)
    EditText mEtForgetVerificationCode;
    @BindView(R.id.tv_forget_send_verification)
    TextView mTvForgetSendVerification;
    @BindView(R.id.btn_next)
    Button mBtnNext;
    @BindView(R.id.ll_forget_phone)
    LinearLayout mLlForgetPhone;
    @BindView(R.id.et_forget_password)
    EditText mEtForgetPassword;
    @BindView(R.id.btn_forget)
    Button mBtnForget;
    @BindView(R.id.ll_forget_password)
    LinearLayout mLlForgetPassword;
    @BindView(R.id.tv_forget_title)
    TextView mTvForgetTitle;
    private int count = 60;
    Handler handler = new Handler();
    String phoneRules = "^1\\d{10}$";
    String passwordRules = "^.{6,16}$";
    Pattern phonePattern = Pattern.compile(phoneRules);
    Pattern passwordPattern = Pattern.compile(passwordRules);
    private String verifyKey,jpushRID,appChannel,appVersion,userId,userName,userIcon,userGender;
    private int channelType,appType,sex;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_forget_password);
        ButterKnife.bind(this);
        jpushRID = JPushInterface.getRegistrationID(mActivity);
        SMSSDK.registerEventHandler(new EventHandle());

        //第三方注册和忘记密码界面的差异
        mTvForgetTitle.setText("绑定手机号");
        mBtnForget.setText("绑定");
        initData();
    }

    private void initData() {
        appType = Constants.APP_TYPE;
        appChannel = WalleChannelReader.getChannel(this);
        appVersion = Utils.getAppVersionName(this);
        Intent intent = getIntent();
        userId = intent.getStringExtra("userId");
        userName = intent.getStringExtra("userName");
        userIcon = intent.getStringExtra("userIcon");
        userGender = intent.getStringExtra("userGender");
        channelType = intent.getIntExtra("channelType", 1);

        if ("m".equals(userGender)) {
            sex = 1;
        } else {
            sex = 2;
        }
    }

    @Override
    public void onBackPressed() {
        if (mLlForgetPassword.getVisibility() == View.VISIBLE) {
            mLlForgetPassword.setVisibility(View.GONE);
            mLlForgetPhone.setVisibility(View.VISIBLE);
            mTvForgetTitle.setText("绑定手机号");
        }
    }

    @Override
    public void onDestroy() {
        EventBus.getDefault().unregister(this);
        super.onDestroy();
        SMSSDK.unregisterAllEventHandler();
    }

    @Override
    protected ForgetPasswordPresenter createPresenter() {
        return new ForgetPasswordPresenter(this);
    }

    @Override
    public void getDataSuccess(LoginBean bean) {
        ZhiZheUtils.saveLoginData(bean);
        EventBus.getDefault().post(new RegisterEvent());
        finish();
    }

    @Override
    public void getPhoneStatusSuccess(boolean isRegistered) {
        if (isRegistered) {
            //手机号确认提示框
            PhoneConfirmDialog confirmDialog = new PhoneConfirmDialog();
            Bundle bundle = new Bundle();
            bundle.putString("phone", mEtForgetPhone.getText().toString());
            confirmDialog.setArguments(bundle);
            confirmDialog.show(mActivity.getFragmentManager(), "");
        } else {
            //手机号未注册提示框
            PhoneUnRegisteredDialog registeredDialog = new PhoneUnRegisteredDialog();
            Bundle bundle = new Bundle();
            bundle.putString("phone", mEtForgetPhone.getText().toString());
            registeredDialog.setArguments(bundle);
            registeredDialog.show(mActivity.getFragmentManager(), "");
        }
    }

    @Override
    public void smsCodeVerificationSuccess(SMSCodeVerificationBean bean) {
        verifyKey = bean.getVerifyKey();
        mLlForgetPhone.setVisibility(View.GONE);
        mLlForgetPassword.setVisibility(View.VISIBLE);
        mTvForgetTitle.setText("设置密码");
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(PhoneConfirmEvent event) {
        //手机号确认成功,发送验证码
        showLoading();
        SMSSDK.getVerificationCode("86", mEtForgetPhone.getText().toString());
    }

    @OnClick(R.id.iv_forget_phone_clear)
    public void onMIvRegisterPhoneClearClicked() {
        mEtForgetPhone.setText("");
    }

    @OnClick(R.id.tv_forget_send_verification)
    public void onMTvRegisterSendVerificationClicked() {
        mPresenter.checkPhoneRegistered(mEtForgetPhone.getText().toString());
    }

    @OnClick(R.id.btn_next)
    public void onMBtnNextClicked() {
        //验证验证码
        mPresenter.smsCodeVerification(mEtForgetPhone.getText().toString()
                , mEtForgetVerificationCode.getText().toString());
    }

    @OnClick(R.id.btn_forget)
    public void onMBtnRegisterClicked() {
        String phone = mEtForgetPhone.getText().toString();
        String password = MD5Utils.md5(mEtForgetPassword.getText().toString());
        mPresenter.channelRegister(channelType, userId, password, userIcon, userName, sex,
                null, phone, verifyKey, jpushRID, appType, appChannel, appVersion);
    }

    @OnTextChanged({R.id.et_forget_phone, R.id.et_forget_verification_code})
    public void onRegisterNextTextChange() {
        mBtnNext.setEnabled(checkPhone() && mEtForgetVerificationCode.length() > 0);
    }

    @OnTextChanged(R.id.et_forget_password)
    public void onRegisterPasswordTextChange() {
        mBtnForget.setEnabled(checkPassword());
    }

    @OnTextChanged(R.id.et_forget_phone)
    public void onRegisterPhoneTextChange() {
        mTvForgetSendVerification.setEnabled(checkPhone());
        if (mEtForgetPhone.length() > 0) {
            mIvForgetPhoneClear.setVisibility(View.VISIBLE);
        } else {
            mIvForgetPhoneClear.setVisibility(View.GONE);
        }
    }

    private boolean checkPhone() {
        if (phonePattern.matcher(mEtForgetPhone.getText().toString()).matches()) {
            return true;
        } else {
            toastShow("手机号格式错误!");
            return false;
        }
    }

    private boolean checkPassword() {
        if (passwordPattern.matcher(mEtForgetPassword.getText().toString()).matches()) {
            return true;
        } else {
            toastShow("密码格式错误!");
            return false;
        }
    }

    Runnable setDjs = new Runnable() {
        @Override
        public void run() {
            mTvForgetSendVerification.setText(count + " S");
            count--;

            if (count < 0) {
                handler.removeCallbacks(setDjs);
                mTvForgetSendVerification.setEnabled(true);
                mTvForgetSendVerification.setText(R.string.re_get_verification);

                count = 60;

                return;
            }

            handler.postDelayed(setDjs, 1000);
        }
    };

    class EventHandle extends EventHandler {
        @Override
        public void afterEvent(final int event, final int result, final Object data) {
            mActivity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    hideLoading();
                    if (result == SMSSDK.RESULT_COMPLETE) {
                        //回调完成
                        if (event == SMSSDK.EVENT_SUBMIT_VERIFICATION_CODE) {
                            //提交验证码成功

                        } else if (event == SMSSDK.EVENT_GET_VERIFICATION_CODE) {
                            //获取验证码成功
                            mTvForgetSendVerification.setEnabled(false);
                            handler.post(setDjs);
                        } else if (event == SMSSDK.EVENT_GET_SUPPORTED_COUNTRIES) {
                            //返回支持发送验证码的国家列表
                        }
                    } else {
                        // 根据服务器返回的网络错误，给toast提示
                        try {
                            Throwable throwable = (Throwable) data;
                            throwable.printStackTrace();
                            JSONObject object = JSON.parseObject(throwable.getMessage());
                            String des = object.getString("detail");//错误描述
                            int status = object.getInteger("status");//错误代码
                            switch (status) {
                                case 457:
                                    toastShow("手机号格式错误!");
                                    break;
                                case 463:
                                case 464:
                                case 465:
                                case 477:
                                case 478:
                                    toastShow("获取验证码次数频繁，请稍后重试");
                                    break;
                                default:
                                    toastShow(des);
                                    break;
                            }
                        } catch (Exception e) {
                            //do something
                        }
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/loginAndRegister/channelRegister/ChannelRegisterActivity.kt
package com.zxcx.zhizhe.ui.loginAndRegister.channelRegister

import android.os.Bundle
import android.os.CountDownTimer
import android.view.View
import butterknife.ButterKnife
import cn.jpush.android.api.JPushInterface
import cn.smssdk.EventHandler
import cn.smssdk.SMSSDK
import com.google.gson.JsonParser
import com.meituan.android.walle.WalleChannelReader
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.event.PhoneConfirmEvent
import com.zxcx.zhizhe.event.StopRegisteredEvent
import com.zxcx.zhizhe.mvpBase.MvpActivity
import com.zxcx.zhizhe.ui.loginAndRegister.login.LoginBean
import com.zxcx.zhizhe.ui.loginAndRegister.login.PhoneConfirmDialog
import com.zxcx.zhizhe.utils.*
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_channel_register.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

class ChannelRegisterActivity : MvpActivity<ChannelRegisterPresenter>(), ChannelRegisterContract.View {

	private var phoneRules = "^1\\d{10}$"
	private var phonePattern = Pattern.compile(phoneRules)
	private var jpushRID: String? = null
	private var userId: String? = null
	private var userName: String? = null
	private var userIcon: String? = null
	private var userGender: String? = null
	private var appChannel: String? = null
	private var appVersion: String? = null
	private var channelType: Int = 0
	private var appType: Int = 0
	private var sex: Int = 0

	val countDownTimer: CountDownTimer = object : CountDownTimer(60000, 1000) {
		override fun onFinish() {
			tv_forget_resend_code.isEnabled = true
			tv_forget_resend_code.text = "重新获取验证码"
		}

		override fun onTick(millisUntilFinished: Long) {
			val value = (millisUntilFinished / 1000).toInt()
			tv_forget_resend_code.text = getString(R.string.tv_login_count_down, value)
		}
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_channel_register)
		ButterKnife.bind(this)
		EventBus.getDefault().register(this)

		jpushRID = JPushInterface.getRegistrationID(mActivity)
		SMSSDK.registerEventHandler(EventHandle())
		initData()
		//延迟弹出软键盘
		Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread())
				.subscribe { Utils.showInputMethod(et_forget_phone) }
	}

	private fun initData() {

		appType = Constants.APP_TYPE
		appChannel = WalleChannelReader.getChannel(this)
		appVersion = Utils.getAppVersionName(this)
		val intent = intent
		userId = intent.getStringExtra("userId")
		userName = intent.getStringExtra("userName")
		userIcon = intent.getStringExtra("userIcon")
		userGender = intent.getStringExtra("userGender")
		channelType = intent.getIntExtra("channelType", 1)

		sex = if ("m" == userGender) 1 else 2
	}

	override fun initStatusBar() {
		//全屏输入法问题
		AndroidBug5497Workaround.assistActivity(this)
	}

	override fun onBackPressed() {
		val stopRegisteredDialog = StopRegisteredDialog()
		stopRegisteredDialog.show(mActivity.fragmentManager, "")
	}

	public override fun onDestroy() {
		EventBus.getDefault().unregister(this)
		super.onDestroy()
		SMSSDK.unregisterAllEventHandler()
	}

	override fun createPresenter(): ChannelRegisterPresenter {
		return ChannelRegisterPresenter(this)
	}

	override fun getDataSuccess(bean: LoginBean) {
		ZhiZheUtils.saveLoginData(bean)
		finish()
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	fun onMessageEvent(event: PhoneConfirmEvent) {
		//手机号确认成功,发送验证码
		showLoading()
		SMSSDK.getVerificationCode("86", et_forget_phone.text.toString())
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	fun onMessageEvent(event: StopRegisteredEvent) {
		finish()
	}

	private fun checkPhone(): Boolean {
		return phonePattern.matcher(et_forget_phone.text.toString()).matches()
	}

	override fun setListener() {
		iv_forget_close.setOnClickListener { onBackPressed() }

		iv_forget_phone_clear.setOnClickListener {
			et_forget_phone.setText("")
			Utils.showInputMethod(et_forget_phone)
		}

		tv_forget_phone.setOnClickListener {
			rl_forget_phone.visibility = View.VISIBLE
			ll_forget_code.visibility = View.GONE
			Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()).subscribe {
				et_forget_phone.requestFocus()
			}
		}

		tv_forget_send_code.setOnClickListener {
			val confirmDialog = PhoneConfirmDialog()
			val bundle = Bundle()
			bundle.putString("phone", et_forget_phone.text.toString())
			confirmDialog.arguments = bundle
			confirmDialog.show(mActivity.fragmentManager, "")
		}

		et_forget_phone.afterTextChanged {
			val isPhone = checkPhone()
			iv_forget_phone_clear.visibility = if (et_forget_phone.length() == 0) View.GONE else View.VISIBLE
			tv_forget_send_code.isEnabled = isPhone
		}

		vci_forget.setOnCompleteListener {
			val jpushRID = JPushInterface.getRegistrationID(mActivity)
			val appType = Constants.APP_TYPE
			val phone = et_forget_phone.text.toString()
			mPresenter.channelRegister(channelType, userId, userIcon, userName, sex,
					null, phone, it, jpushRID, appType, appChannel, appVersion)
		}
	}

	internal inner class EventHandle : EventHandler() {
		override fun afterEvent(event: Int, result: Int, data: Any) {
			mActivity.runOnUiThread {
				hideLoading()
				if (result == SMSSDK.RESULT_COMPLETE) {
					//回调完成
					when (event) {
						SMSSDK.EVENT_SUBMIT_VERIFICATION_CODE -> { /*提交验证码成功*/
						}
						SMSSDK.EVENT_GET_VERIFICATION_CODE -> {
							/*获取验证码成功*/
							toastShow(R.string.send_verification_code_success)
							tv_forget_phone.text = et_forget_phone.text.toString()
							rl_forget_phone.visibility = View.GONE
							ll_forget_code.visibility = View.VISIBLE
							tv_forget_resend_code.isEnabled = false
							Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()).subscribe {
								vci_forget.focus()
							}
							countDownTimer.start()
						}
						SMSSDK.EVENT_GET_SUPPORTED_COUNTRIES -> {/*返回支持发送验证码的国家列表*/
						}
					}
				} else {
					// 根据服务器返回的网络错误，给toast提示
					try {
						val throwable = data as Throwable
						throwable.printStackTrace()
						val `object` = JsonParser().parse(throwable.message).asJsonObject
						val des = `object`.get("detail").asString//错误描述
						val status = `object`.get("status").asInt//错误代码
						when (status) {
							457 -> toastError("手机号格式错误!")
							463, 464, 465, 477, 478 -> toastError("获取验证码次数频繁，请稍后重试")
							else -> toastError(des)
						}
					} catch (e: Exception) {
						//do something
					}

				}
			}
		}
	}
}
